### PR TITLE
added afterLogout trigger

### PR DESCRIPTION
**Pull request by mtrezza**
Analogous to the existing `beforeLogin` trigger, this `afterLogout` trigger can be used to perform certain clean-up actions upon user logout.

The `afterLogout` trigger contains the `_Session` object that was deleted. Usually, a session object contains a pointer to a `_User` object with which the session is associated. With that, actions can be taken for the specific user that logged out.

Caveat:
The `afterLogout` is only triggered when a session object was deleted upon logout. If a user logs out but there is not session object to be deleted, the `afterLogout` trigger is not called. This is by design, because without a session object to determine _who_ logged out, the `afterLogout` trigger does not seem useful.

### CODE REVIEW

**Comment by mtrezza** [comment]
Notes:
- tests included
- `afterLogout` is a trigger on the `_Session` class, so the trigger validation had to be changed. Before, any trigger on the `_Session` class was forbidden, now only the `afterLogout` trigger is allowed for the `_Session` class.

File: src/RestWrite.js [hunk 1]
```diff
@@ -321,6 +321,33 @@ RestWrite.prototype.runBeforeLoginTrigger = async function(userData) {
   );
 };
 
+RestWrite.prototype.runAfterLogoutTrigger = async function(sessionData) {
```
Comment (dplewis) [inline src/RestWrite.js]: Is this being used?

File: spec/CloudCode.spec.js [hunk 2]
```diff
@@ -2436,6 +2456,19 @@ describe('beforeLogin hook', () => {
     done();
   });
 
+  it('should trigger afterLogout hook on logout', async done => {
+    let hit = 0;
+    Parse.Cloud.afterLogout(() => {
+      hit++;
```
Comment (dplewis) [inline spec/CloudCode.spec.js]: After Logout does return a request object. Can you test it?

File: src/triggers.js [hunk 3]
```diff
@@ -32,7 +33,8 @@ const baseStore = function() {
 };
 
 function validateClassNameForTriggers(className, type) {
-  const restrictedClassNames = ['_Session'];
+  // Contains classes which do not allow triggers of any kind
+  const restrictedClassNames = [];
   if (restrictedClassNames.indexOf(className) != -1) {
```
Comment (dplewis) [inline src/triggers.js]: This can be deleted no?

Comment (mtrezza) [reply src/triggers.js]: Yes, I thought I leave it as placeholder, but who knows if we ever need it again. Will remove it.

File: src/RestWrite.js (hunk 1 above)
Comment (mtrezza) [reply src/RestWrite.js]: nope, removed

File: spec/CloudCode.spec.js (hunk 2 above)
Comment (mtrezza) [reply spec/CloudCode.spec.js]: added to tests

**Comment by dplewis** [comment]
> The afterLogout trigger contains the _Session object that was deleted. 

Is this true?

**Comment by mtrezza** [comment]
@dplewis Yes, the session object is deleted from the DB and returned here. If there was any session object to delete.

See `this._runAfterLogoutTrigger(req, records.results[0]);`.

**Comment by dplewis** [comment]
Would I get the user?

**Comment by mtrezza** [comment]
@dplewis Yes, you can get the user, if there was a user associated with the session.

**Comment by mtrezza** [comment]
@dplewis I just noticed that in the tests I do not get the user from the session. The session object only looks like this:
`ParseSession { _objCount: 5, className: '_Session', id: 'IigxK0tazS' }`. 

There seem to be fields missing, do you know why?

**Comment by dplewis** [comment]
It is an object I assume you can do req.object.get(“user”) and fetch the user. Can you add it to the tests.

**Comment by mtrezza** [comment]
Oh you are right, that works. I will add this to the tests.

**Comment by mtrezza** [comment]
Thanks all for the fast review 🚀

**Comment by vince1995** [comment]
Sorry for pulling up this old thread but is there a reason for`now only the afterLogout trigger is allowed for the _Session class`? Why not all the other triggers?

**Pull request merged**